File: backend/app/schemas/alerts.py
from datetime import datetime
import re
from typing import Literal
from urllib.parse import urlparse

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


Severity = Literal["Low", "Medium", "High", "Critical"]
SourceHealthAlertMode = Literal["new_error", "every_error"]


class AlertPreferenceWrite(BaseModel):
    enabled: bool = False
    telegram_enabled: bool = False
    telegram_bot_token: str = Field(default="", max_length=160)
    clear_telegram_bot_token: bool = False
    telegram_chat_id: str = Field(default="", max_length=80)
    discord_enabled: bool = False
    discord_webhook_url: str = Field(default="", max_length=600)
    clear_discord_webhook: bool = False
    notify_news: bool = True
    notify_cve: bool = True
    notify_asset_exposure: bool = True
    notify_ioc: bool = False
    notify_source_health: bool = True
    minimum_severity: Severity = "High"
    cve_minimum_severity: Severity = "High"
    asset_exposure_minimum_severity: Severity = "High"
    ioc_minimum_severity: Severity = "High"
    source_health_alert_mode: SourceHealthAlertMode = "new_error"

    @field_validator("telegram_chat_id")
    @classmethod
    def validate_chat_id(cls, value: str) -> str:
        value = value.strip()
        if value and not value.removeprefix("-").isdigit():
            raise ValueError("Telegram chat ID harus berupa angka")
        return value

    @field_validator("telegram_bot_token")
    @classmethod
    def validate_telegram_bot_token(cls, value: str) -> str:
        value = value.strip()
        if value and not re.fullmatch(r"\d{5,20}:[A-Za-z0-9_-]{20,}", value):
            raise ValueError("Telegram bot token tidak valid")
        return value

    @field_validator("discord_webhook_url")
    @classmethod
    def validate_discord_webhook(cls, value: str) -> str:
        value = value.strip()
        if not value:
            return value
        parsed = urlparse(value)
        if parsed.scheme != "https" or parsed.hostname not in {"discord.com", "discordapp.com"}:
            raise ValueError("Gunakan URL webhook Discord resmi dengan HTTPS")
        if not parsed.path.startswith("/api/webhooks/"):
            raise ValueError("URL webhook Discord tidak valid")
        return value

    @model_validator(mode="after")
    def validate_enabled_channels(self) -> "AlertPreferenceWrite":
        if self.enabled and not (self.telegram_enabled or self.discord_enabled):
            raise ValueError("Aktifkan minimal satu channel")
        if self.telegram_enabled and not self.telegram_chat_id:
            raise ValueError("Telegram chat ID wajib diisi")
        return self


class AlertPreferenceOut(BaseModel):
    enabled: bool
    telegram_enabled: bool
    telegram_bot_configured: bool
    telegram_chat_id: str
    discord_enabled: bool
    discord_webhook_configured: bool
    discord_webhook_hint: str
    notify_news: bool
    notify_cve: bool
    notify_asset_exposure: bool
    notify_ioc: bool
    notify_source_health: bool
    minimum_severity: Severity
    cve_minimum_severity: Severity
    asset_exposure_minimum_severity: Severity
    ioc_minimum_severity: Severity
    source_health_alert_mode: SourceHealthAlertMode
    updated_at: datetime | None


class AlertDeliveryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    event_type: str
    channel: str
    title: str
    severity: str
    status: str
    retry_count: int = 0
    error: str
    created_at: datetime
    sent_at: datetime | None


class AlertTestRequest(BaseModel):
    channel: Literal["telegram", "discord"]
    telegram_bot_token: str = Field(default="", max_length=160)
    telegram_chat_id: str = Field(default="", max_length=80)
    discord_webhook_url: str = Field(default="", max_length=600)

    @field_validator("telegram_bot_token")
    @classmethod
    def validate_test_telegram_bot_token(cls, value: str) -> str:
        return AlertPreferenceWrite.validate_telegram_bot_token(value)

    @field_validator("telegram_chat_id")
    @classmethod
    def validate_test_chat_id(cls, value: str) -> str:
        return AlertPreferenceWrite.validate_chat_id(value)

    @field_validator("discord_webhook_url")
    @classmethod
    def validate_test_discord_webhook(cls, value: str) -> str:
        return AlertPreferenceWrite.validate_discord_webhook(value)


class AlertTestResponse(BaseModel):
    status: Literal["sent"]
    channel: Literal["telegram", "discord"]

File: frontend/src/lib/alerts.ts
export type AlertSeverity = "Low" | "Medium" | "High" | "Critical";

export type AlertPreference = {
  enabled: boolean;
  telegram_enabled: boolean;
  telegram_bot_configured: boolean;
  telegram_chat_id: string;
  discord_enabled: boolean;
  discord_webhook_configured: boolean;
  discord_webhook_hint: string;
  notify_news: boolean;
  notify_cve: boolean;
  notify_asset_exposure: boolean;
  notify_ioc: boolean;
  notify_source_health: boolean;
  minimum_severity: AlertSeverity;
  cve_minimum_severity: AlertSeverity;
  asset_exposure_minimum_severity: AlertSeverity;
  ioc_minimum_severity: AlertSeverity;
  source_health_alert_mode: "new_error" | "every_error";
  updated_at: string | null;
};

export type AlertPreferenceInput = AlertPreference & {
  telegram_bot_token: string;
  clear_telegram_bot_token: boolean;
  discord_webhook_url: string;
  clear_discord_webhook: boolean;
};

export type AlertDelivery = {
  id: number;
  event_type: string;
  channel: "telegram" | "discord" | string;
  title: string;
  severity: string;
  status: "sent" | "failed" | string;
  retry_count: number;
  error: string;
  created_at: string;
  sent_at: string | null;
};

async function request<T>(url: string, options: RequestInit = {}): Promise<T> {
  const response = await fetch(url, {
    credentials: "same-origin",
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...options.headers
    }
  });
  if (!response.ok) {
    const body = await response.json().catch(() => ({ detail: "Request failed" }));
    throw new Error(typeof body.detail === "string" ? body.detail : "Request failed");
  }
  return response.json();
}

export function fetchAlertPreferences() {
  return request<AlertPreference>("/api/alerts/preferences");
}

export function updateAlertPreferences(payload: AlertPreferenceInput, csrfToken: string) {
  const {
    discord_webhook_configured,
    discord_webhook_hint,
    telegram_bot_configured,
    updated_at,
    ...writePayload
  } = payload;
  void discord_webhook_configured;
  void discord_webhook_hint;
  void telegram_bot_configured;
  void updated_at;
  return request<AlertPreference>("/api/alerts/preferences", {
    method: "PUT",
    headers: { "X-CSRF-Token": csrfToken },
    body: JSON.stringify(writePayload)
  });
}

export function testAlertChannel(
  channel: "telegram" | "discord",
  csrfToken: string,
  payload: Partial<Pick<AlertPreferenceInput, "telegram_bot_token" | "telegram_chat_id" | "discord_webhook_url">> = {}
) {
  return request<{ status: "sent"; channel: "telegram" | "discord" }>("/api/alerts/test", {
    method: "POST",
    headers: { "X-CSRF-Token": csrfToken },
    body: JSON.stringify({ channel, ...payload })
  });
}

export function fetchAlertDeliveries() {
  return request<AlertDelivery[]>("/api/alerts/deliveries");
}
